File: tools/journey-logic.py
#!/usr/bin/env python3
"""journey-logic.py — PROOF A (bare logic) of the onboarding journey test cycle.

    python3 tools/journey-logic.py --selftest        # the mutation suite — beat 2's own precondition
    python3 tools/journey-logic.py                   # the 13-path table against QA
    python3 tools/journey-logic.py --json

Design per `.plans/2026-09-05-journey-test-cycle-PROPOSAL.md` §2 beats 1-3 and §6 (PROPOSED, unruled).

Three properties this tool exists to have, each because something already went wrong without it:
  · The identity marker is DERIVED from the working tree (the <title> text and the id="s…" roster
    parsed out of onboarding/index.html at run time), never typed. A hand-typed marker rotted within
    eight hours: the 09-05 cascade proposal says assert `Your place`; the document says `My Home`.
  · It FAILS CLOSED on the target before it asserts anything — a wrong build, a non-qa env, or lab.
    Every path on lab returns the same document with HTTP 200, including /nonexistent.
  · It is proven able to FAIL. --selftest plants five defects on a scratch copy and requires each to
    flip one named assertion red. A suite that has only ever passed has proven nothing.

It writes NOTHING. The document is served by route interception, so a run leaves no residue in KV,
no fixture row and no tracked file.

⛔ THE BOUNDARY, AND IT IS NOT A SMALL ONE (measured by practice-steward, 2026-09-05):
`identity_from_tree()` derives the screen roster FROM THE DOCUMENT IT IS ABOUT TO TEST. So the
completeness predicate is the document's own screen list, and **"this journey is missing its opening
stages" is structurally unreachable by this tool.** On 2026-09-05 it passed 15/15 paths and 5/5
mutations over a journey with no account step, no hub and no name ask anywhere — every one of which
Paul then named as missing within the hour. Nothing here was wrong; there was no upstream declaration
to compare against.

⭐ So a green run means: *every path this document contains resolves as stated.* It does NOT mean the
journey is complete, and no mutation of this document can make it say otherwise. The oracle for
completeness has to come from OUTSIDE the artifact — a declared stage list. Until one exists, read a
green run as coverage of what was built, never as coverage of what a person needs.
"""
import argparse, json, os, re, shutil, subprocess, sys, tempfile, urllib.request

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import qa_access

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
DOC = os.path.join(ROOT, "onboarding", "index.html")
UA = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)"

QA_ORIGIN = "https://fernwood-qa.pages.dev"
QA_WORKER = "https://fernwood-qa.paul-kirschenbauer.workers.dev"
DOC_PATH = "/onboarding/"


def get(url, timeout=30):
    req = urllib.request.Request(url, headers={**qa_access.headers(url), "User-Agent": UA})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return r.status, r.read()


# ── the DERIVED identity marker — no fact a human types is state ────────────────────────────────
def identity_from_tree(path=DOC):
    src = open(path, encoding="utf-8").read()
    m = re.search(r"<title>([^<]*)</title>", src)
    if not m:
        raise SystemExit("journey-logic: onboarding/index.html has no <title> — nothing to assert on")
    screens = re.findall(r'id="(s-nolink|s-wait|s\d)"', src)
    roster = sorted(set(screens))
    steps = [s for s in roster if re.fullmatch(r"s\d", s)]
    if not steps:
        raise SystemExit("journey-logic: no id=\"s<n>\" screens found — the document shape changed")
    return {"title": m.group(1).strip(), "screens": roster, "steps": steps}


# ── beat 1 · FIX THE TARGET, fail closed three ways ─────────────────────────────────────────────
def fix_target(intended_sha, allow_stale=False):
    out = {}
    # ⛔ GATE 1 RUNS ON QA, NEVER ON LAB — and the reason recorded here until 2026-09-05 was WRONG in a
    # way that invited the wrong fix. It said "every path there returns the same 200 document". That IS
    # true of lab: /definitely-not-a-real-path.json returns 200 with the root document. But it is
    # EQUALLY true of QA — the SPA fallback is a Cloudflare Pages property, not a lab property — so the
    # stated reason does not discriminate between the two, and anyone who checked it could "fix" this
    # by pointing at QA and believe they had solved something. Measured on both origins.
    #
    # The reasons that actually discriminate:
    #   1. QA has its OWN estate (est-qa0001). Lab shared PRODUCTION's estate id until 2026-09-05, so
    #      lab records were byte-identical in prefix to prod's and separable only by namespace. Now
    #      est-lab0001 — but "not prod's id" is not the same as "the gate-1 estate".
    #   2. QA's build stamp is written into the Pages export by deploy-worker-qa.yml and is CURRENT.
    #      Lab is deployed by hand, qa-build.json is never a tracked file, so lab's stamp is whatever
    #      some earlier CI run left behind. On 2026-09-05 lab served HEAD 6d61fc0 while its own stamp
    #      claimed 99cc226 — an origin that cannot say which build it is serving cannot anchor a gate.
    if "lab" in QA_ORIGIN:
        raise SystemExit(
            "journey-logic: ⛔ never lab — gate 1 runs on QA. Lab is hand-deployed and its "
            "qa-build.json is left over from CI, so it cannot say which build it is serving. "
            "(Note: the 200-for-any-path behaviour is a Pages property QA shares — that is NOT "
            "the reason, and pointing this at QA to dodge the check fixes nothing.)")
    try:
        st, body = get(QA_ORIGIN + "/qa-build.json")
        build = json.loads(body)
    except Exception as e:
        raise SystemExit("journey-logic: cannot read qa-build.json (%s) — target unresolvable, refusing" % e)
    out["servedSha"] = build.get("sha")
    out["servedSubj"] = build.get("subject")
    out["origin"] = QA_ORIGIN
    if out["servedSha"] != intended_sha:
        msg = ("journey-logic: ⛔ QA is serving %s but you intend to walk %s — you would be walking a "
               "different build" % ((out["servedSha"] or "?")[:7], intended_sha[:7]))
        if not allow_stale:
            raise SystemExit(msg + "\n  (--allow-stale to walk what QA actually serves, recorded as such)")
        out["staleTargetAccepted"] = msg
    try:
        st, body = get(QA_WORKER + "/health")
        health = json.loads(body)
    except Exception as e:
        raise SystemExit("journey-logic: cannot read the QA Worker /health (%s) — refusing" % e)
    if health.get("env") != "qa" or health.get("kv_canary") != "qa":
        raise SystemExit("journey-logic: ⛔ /health says env=%r kv_canary=%r — not QA. Nothing run."
                         % (health.get("env"), health.get("kv_canary")))
    out["env"] = health.get("env")
    out["kvCanary"] = health.get("kv_canary")
    out["estateId"] = health.get("estateId")
    return out


def chromium_path():
    if os.environ.get("QA_WALK_CHROMIUM"):
        return os.environ["QA_WALK_CHROMIUM"]
    return None


def node_modules_dir():
    import glob
    cands = glob.glob(os.path.expanduser("~/.npm/_npx/*/node_modules/playwright"))
    if not cands:
        raise SystemExit("journey-logic: no cached playwright under ~/.npm/_npx — cannot run")
    return os.path.dirname(cands[0])


def run_paths(doc_body, login_bytes, identity, settle=900):
    cfg = {
        "origin": QA_ORIGIN, "docPath": DOC_PATH, "docBody": doc_body,
        "loginBytes": login_bytes, "identity": identity,
        "screens": identity["screens"], "grant": "qa-synth-1-fixture-token",
        "settleMs": settle,
    }
    env = dict(os.environ)
    env["NODE_PATH"] = node_modules_dir()
    p = subprocess.run(["node", os.path.join(HERE, "journey-logic.js"), json.dumps(cfg)],
                       capture_output=True, text=True, env=env, timeout=600)
    if p.returncode != 0 or not p.stdout.strip():
        raise SystemExit("journey-logic: the driver failed\n" + (p.stderr or "")[-2000:])
    return json.loads(p.stdout)["results"]


# ── §6a · the mutation suite — five planted defects, each must flip one named path red ───────────
MUTATIONS = [
    ("m-guard", 'if (!grant || !WORKER) { show("s-nolink"); return; }', "if (false) { show(\"s-nolink\"); return; }", 2),
    ("m-worker-hardcode", "? PAGES_WORKERS[label] : null;", ": PAGES_WORKERS[\"fernwood-qa\"];", 5),
    # ⚠️ anchored on postAnswer()'s fingerprint since the 2026-09-05 refactor moved it out of go2.
    # The suite REFUSED to score this green when the anchor moved, which is the property that matters.
    ("m-fp-grant-only", 'var gv = (read(K_GRANT) || "x") + "\\u0000" + note;', 'var gv = (read(K_GRANT) || "x");', 9),
    ("m-stored-zero", "if (res && res.stored === 0 && !res.duplicate) return Promise.reject(\"not-stored\");",
     "if (false) return Promise.reject(\"not-stored\");", 10),
    # naming became the first step 2026-09-05; the mutation was written the same hour, because a path
    # nothing can flip red is a path that proves nothing about the screen it names.
    ("m-name-optional", 'if (!nm) {', 'if (false) {', 14),
]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--selftest", action="store_true", help="plant five defects; each must flip one path red")
    ap.add_argument("--json", action="store_true")
    ap.add_argument("--allow-stale", action="store_true", help="walk what QA serves even if it is not HEAD")
    a = ap.parse_args()

    identity = identity_from_tree()
    doc_body = open(DOC, encoding="utf-8").read()
    head = subprocess.run(["git", "-C", ROOT, "rev-parse", "HEAD"], capture_output=True, text=True).stdout.strip()

    # the Access login page — real bytes, fetched tokenless, for m-wrong-doc / path 13
    req = urllib.request.Request(QA_ORIGIN + DOC_PATH, headers={"User-Agent": UA})
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            login_bytes = r.read().decode("utf-8", "replace")
    except Exception as e:
        raise SystemExit("journey-logic: could not fetch the Access login page for path 13 (%s)" % e)
    if "<title>My Home</title>" in login_bytes:
        raise SystemExit("journey-logic: ⛔ the tokenless fetch returned the PRODUCT — Access is not gating QA. "
                         "path 13 would be vacuous; refusing to run.")

    target = fix_target(head, allow_stale=a.allow_stale)

    if a.selftest:
        return selftest(doc_body, login_bytes, identity, target)

    results = run_paths(doc_body, login_bytes, identity)
    return report(results, identity, target, a.json, head)


def selftest(doc_body, login_bytes, identity, target):
    print("journey-logic --selftest — five planted defects, each must flip one named path red\n")
    base = run_paths(doc_body, login_bytes, identity)
    base_bad = [r for r in base if not r["ok"]]
    ok_clean = not base_bad
    print("  %s the UNMUTATED run is clean (%d/%d paths)"
          % ("✅" if ok_clean else "🔴", len(base) - len(base_bad), len(base)))
    for r in base_bad:
        print("       path %s — %s · %s" % (r["id"], r["name"], r["detail"]))

    caught, ran = [], []
    for mid, find, repl, path_id in MUTATIONS:
        ran.append(mid)
        if find not in doc_body:
            print("  🔴 %-20s ANCHOR NOT FOUND — the document changed; this mutation proves nothing" % mid)
            continue
        mutated = doc_body.replace(find, repl, 1)
        res = run_paths(mutated, login_bytes, identity)
        row = [r for r in res if r["id"] == path_id]
        flipped = row and not all(r["ok"] for r in row)
        base_row_ok = all(r["ok"] for r in base if r["id"] == path_id)
        good = flipped and base_row_ok
        if good:
            caught.append(mid)
        print("  %s %-20s → path %-2s %s" % ("✅" if good else "🔴", mid, path_id,
              "caught" if good else ("NOT caught — the assertion set has a hole here"
                                     if base_row_ok else "path %s was already red; inconclusive" % path_id)))

    # m-wrong-doc is structural: serve the login bytes as the document
    res = run_paths(login_bytes, login_bytes, identity)
    row13 = [r for r in res if r["id"] == 13]
    wrongdoc_ok = bool(row13) and row13[0]["ok"]
    ran.append("m-wrong-doc")
    if wrongdoc_ok:
        caught.append("m-wrong-doc")
    print("  %s %-20s → path 13 %s" % ("✅" if wrongdoc_ok else "🔴", "m-wrong-doc",
          "caught (the runner refuses a 200-status page that is not the product)"
          if wrongdoc_ok else "NOT caught — the runner cannot tell the product from a login page"))

    green = ok_clean and len(caught) == len(ran)
    print("\n%s selftest: %d/%d mutations caught, unmutated run %s"
          % ("✅" if green else "🔴", len(caught), len(ran), "clean" if ok_clean else "NOT clean"))
    print("   mutations: %s" % json.dumps({"ran": len(ran), "caught": len(caught), "ids": ran}))
    return 0 if green else 1


def report(results, identity, target, as_json, head):
    if as_json:
        print(json.dumps({"head": head, "target": target, "identity": identity, "results": results}, indent=2))
    bad = [r for r in results if not r["ok"]]
    print("journey-logic — PROOF A · bare logic")
    print("  target: %s @ %s · env=%s · estate=%s" % (target["origin"], (target.get("servedSha") or "?")[:7],
                                                      target.get("env"), target.get("estateId")))
    print("  identity (derived from the tree): title=%r screens=%s" % (identity["title"], ",".join(identity["screens"])))
    if target.get("staleTargetAccepted"):
        print("  ⚠️  %s" % target["staleTargetAccepted"])
    print()
    for r in results:
        print("  %s path %-2s %s" % ("✅" if r["ok"] else "🔴", r["id"], r["name"]))
        if r.get("detail"):
            print("        %s" % r["detail"])
        if r.get("note"):
            print("        ⚠️  %s" % r["note"])
    print()
    print("%s %d/%d paths" % ("✅" if not bad else "🔴", len(results) - len(bad), len(results)))
    return 0 if not bad else 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/journey-logic.js
// journey-logic.js — the bare-logic path table for the onboarding journey, forced state by state.
// Driven by tools/journey-logic.py; never run directly. Config arrives as JSON on argv[2].
//
// The document under test is ALWAYS served by interception, so a run is byte-exact and leaves no
// residue. The Worker's responses are controlled per path. The identity marker is DERIVED by the
// Python side from the working tree and passed in — no marker is typed here.
const { chromium } = require('playwright');
const cfg = JSON.parse(process.argv[2]);

const ORIGIN = cfg.origin;                 // e.g. https://fernwood-qa.pages.dev
const DOC_URL = ORIGIN + cfg.docPath;      // e.g. /onboarding/

// ── helpers ────────────────────────────────────────────────────────────────
const CORS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type, X-Grant, X-Tate-Token',
  'Access-Control-Max-Age': '600',
};
const results = [];
function record(id, name, ok, detail, extra) {
  results.push(Object.assign({ id, name, ok: !!ok, detail: detail || '' }, extra || {}));
}

// Build a context whose document request is fulfilled with `body`, and whose Worker calls are
// answered by `api` (a map of substring -> {status, json} | 'abort' | 'live').
async function openPage(browser, { body, api, storage, url, host }) {
  const ctx = await browser.newContext({ viewport: { width: 414, height: 848 } });
  ctx.setDefaultTimeout(5000);
  const target = url || DOC_URL;
  const posts = [];
  await ctx.route('**/*', async (route) => {
    const u = route.request().url();
    // the document itself (and anything on the Pages origin that is not an API call)
    if (u.startsWith(target.split('?')[0]) || u === target) {
      return route.fulfill({ status: 200, contentType: 'text/html; charset=utf-8', body });
    }
    for (const key of Object.keys(api || {})) {
      if (u.includes(key)) {
        const rule = api[key];
        if (rule === 'abort') return route.abort('failed');
        // The page lives on *.pages.dev and calls the Worker on *.workers.dev — CROSS-ORIGIN. Both
        // `X-Grant` and `Content-Type: application/json` are non-simple, so the browser sends a
        // preflight first. A fulfilled response without CORS headers is BLOCKED by the browser, the
        // fetch rejects, and the page takes its offline branch — i.e. every path silently resolves to
        // s-nolink and the runner reports the product broken when it is the harness that is.
        if (route.request().method() === 'OPTIONS') {
          return route.fulfill({ status: 204, headers: CORS, body: '' });
        }
        if (route.request().method() === 'POST') {
          try { posts.push({ url: u, body: JSON.parse(route.request().postData() || '{}') }); }
          catch (e) { posts.push({ url: u, body: null }); }
        }
        return route.fulfill({ status: rule.status,
                               headers: Object.assign({ 'Content-Type': 'application/json' }, CORS),
                               body: JSON.stringify(rule.json === undefined ? {} : rule.json) });
      }
    }
    // anything else on the page (there should be none — the page makes no third-party request)
    return route.fulfill({ status: 404, body: '' });
  });
  const page = await ctx.newPage();
  const errors = [];
  page.on('pageerror', (e) => errors.push(String(e && e.message || e)));
  page.__errors = errors;
  if (storage) {
    // seed localStorage on the right origin before the document runs
    await page.addInitScript((s) => {
      try { for (const k of Object.keys(s)) localStorage.setItem(k, s[k]); } catch (e) {}
    }, storage);
  }
  await page.goto(target, { waitUntil: 'load', timeout: 45000 });
  await page.waitForTimeout(cfg.settleMs || 900);
  page.__posts = posts;
  page.__ctx = ctx;
  return page;
}

// which of the six screens is visible
async function visible(page) {
  return page.evaluate((ids) => {
    const shown = ids.filter((id) => {
      const el = document.getElementById(id);
      return el && !el.hidden;
    });
    return shown;
  }, cfg.screens);
}
async function nolinkHTML(page) {
  return page.evaluate(() => {
    const el = document.getElementById('s-nolink');
    return el ? el.innerHTML.replace(/\s+/g, ' ').trim() : null;
  });
}

// ── the run ────────────────────────────────────────────────────────────────
(async () => {
  const exe = process.env.QA_WALK_CHROMIUM || null;
  const browser = await chromium.launch(exe ? { executablePath: exe } : {});
  const DOC = cfg.docBody;
  const OK_WHOAMI = { status: 200, json: { personId: 'p-qa-fixture-a', estateId: 'est-qa0001', capability: 'member' } };
  const NOT_FOUND = { status: 404, json: { error: 'not found' } };
  const api = (whoami, feedback) => {
    const m = {};
    if (whoami) m['/api/grant/whoami'] = whoami;
    if (feedback) m['/api/feedback'] = feedback;
    return m;
  };

  try {
    // ─── PATH 13 FIRST — document identity. If the runner cannot tell the product from a
    // 200-status page that is not the product, nothing else it reports means anything.
    {
      const p = await openPage(browser, { body: cfg.loginBytes, api: api(OK_WHOAMI) });
      const title = await p.title();
      const ids = await p.evaluate((s) => s.map((id) => !!document.getElementById(id)), cfg.screens);
      const isProduct = title === cfg.identity.title && ids.every(Boolean);
      record(13, 'wrong document (Access login, HTTP 200) is REFUSED', !isProduct,
             isProduct ? 'the runner accepted a non-product page' : `saw title="${title}"`);
      await p.__ctx.close();
    }

    // ⛔ THE GATE: if the document under test is not the product, STOP. Asserting paths 1-12 against
    // a page that has none of the elements is not a red run — it is a runner interrogating the wrong
    // document, which is the exact confusion path 13 exists to prevent. (It also hangs: every
    // fill()/click() waits out its timeout on a selector that will never appear.)
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI) });
      const title = await p.title();
      const ids = await p.evaluate((s) => s.map((id) => !!document.getElementById(id)), cfg.screens);
      await p.__ctx.close();
      if (!(title === cfg.identity.title && ids.every(Boolean))) {
        record(0, 'the document under test IS the product', false,
               `served title="${title}" but the tree says "${cfg.identity.title}" — paths 1-12 not run`);
        await browser.close();
        console.log(JSON.stringify({ results, abortedOnIdentity: true }, null, 2));
        return;
      }
    }

    // ─── PATH 1 — the link carries the credential, and the credential leaves the address bar
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI), url: DOC_URL + '?g=' + cfg.grant });
      const shown = await visible(p);
      const search = await p.evaluate(() => location.search);
      const stored = await p.evaluate(() => localStorage.getItem('fw-grant'));
      record(1, 'valid ?g on a known host → s1, ?g stripped, and nothing thrown',
             shown.join() === 's1' && !/[?&]g=/.test(search) && stored === cfg.grant && p.__errors.length === 0,
             `screen=${shown.join()||'none'} search="${search}" stored=${stored ? 'yes' : 'no'} scriptErrors=${JSON.stringify(p.__errors)}`);
      await p.__ctx.close();
    }

    // ─── PATH 2 — no credential at all
    let nolink2 = null;
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI) });
      const shown = await visible(p);
      nolink2 = await nolinkHTML(p);
      record(2, 'no g, empty storage → s-nolink', shown.join() === 's-nolink', `screen=${shown.join()||'none'}`);
      await p.__ctx.close();
    }

    // ─── PATH 3 — an unknown/revoked credential must look IDENTICAL to path 2
    {
      const p = await openPage(browser, { body: DOC, api: api(NOT_FOUND), url: DOC_URL + '?g=bogus-not-a-grant' });
      const shown = await visible(p);
      const html = await nolinkHTML(p);
      record(3, 'unknown/revoked grant → s-nolink, byte-identical to path 2',
             shown.join() === 's-nolink' && html === nolink2,
             shown.join() !== 's-nolink' ? `screen=${shown.join()||'none'}`
               : (html === nolink2 ? 'identical' : 'COPY DIFFERS from path 2 — the failure screen leaks which case it was'));
      await p.__ctx.close();
    }

    // ─── PATH 4 — offline is NOT a bad link and must not be dressed as one
    {
      const p = await openPage(browser, { body: DOC, api: api('abort'), url: DOC_URL + '?g=' + cfg.grant });
      const shown = await visible(p);
      const html = await nolinkHTML(p);
      const differs = html !== nolink2;
      const saysConn = /connection|wi-?fi/i.test(html || '');
      // ⛔ 'the offline copy appeared' is satisfied by the very catastrophe it must exclude: when
      // step() threw on every call (09-05), EVERY path showed this copy and this row went green.
      // A genuine offline must ALSO be quiet — no script fault behind it.
      const quiet = p.__errors.length === 0;
      record(4, 'whoami unreachable → the OFFLINE copy, and NOTHING was thrown behind it',
             shown.join() === 's-nolink' && differs && saysConn && quiet,
             `screen=${shown.join()||'none'} differsFromBadLink=${differs} mentionsConnection=${saysConn} scriptErrors=${JSON.stringify(p.__errors)}`);
      await p.__ctx.close();
    }

    // ─── PATH 5 — an unknown Pages host must fail closed (WORKER resolves null)
    {
      const bogusOrigin = 'https://fernwood-not-a-real-label.pages.dev';
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI),
                                          url: bogusOrigin + cfg.docPath + '?g=' + cfg.grant });
      const shown = await visible(p);
      record(5, 'unknown Pages host → s-nolink (WORKER null, fail-closed)',
             shown.join() === 's-nolink', `screen=${shown.join()||'none'}`);
      await p.__ctx.close();
    }

    // ─── PATH 6 — the resume table, including the deliberate 3 → s4
    for (const [step, want] of [['2', 's2'], ['3', 's4'], ['4', 's4']]) {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI),
                                          storage: { 'fw-grant': cfg.grant, 'fw-onboard-step': step } });
      const shown = await visible(p);
      record(6, `resume: fw-onboard-step=${step} → ${want}`, shown.join() === want,
             `screen=${shown.join()||'none'}`);
      await p.__ctx.close();
    }

    // ─── PATH 7 — a missing required field names WHAT is missing and never scolds
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI, { status: 200, json: { stored: 1 } }),
                                          storage: { 'fw-grant': cfg.grant, 'fw-onboard-step': '2' } });
      await p.fill('#a1', '123 Example Rd');          // city, state, zip deliberately left empty
      await p.click('#go2');
      await p.waitForTimeout(300);
      const shown = await visible(p);
      const trouble = await p.evaluate(() => {
        const t = document.getElementById('trouble');
        return { hidden: t.hidden, text: (t.textContent || '').trim() };
      });
      const focused = await p.evaluate(() => document.activeElement && document.activeElement.id);
      const names = /city/i.test(trouble.text) && /state/i.test(trouble.text) && /ZIP/i.test(trouble.text);
      record(7, 'submit with gaps → stays on s2, names what is missing, focuses the first gap',
             shown.join() === 's2' && !trouble.hidden && names && focused === 'city',
             `screen=${shown.join()||'none'} trouble="${trouble.text}" focus=${focused}`);
      await p.__ctx.close();
    }

    // ─── PATHS 8 + 9 — the fingerprint, across THREE fresh sessions.
    // Replay must dedup to the SAME id; a correction must produce a DIFFERENT one. Doing this in one
    // page means clicking #go2 while it is still disabled with "Saving…" — a harness fault, not a
    // finding — so each submit gets its own context, which is also the realistic shape (she comes
    // back later and fixes it).
    {
      const submit = async (line1, feedbackJson) => {
        const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI, { status: 200, json: feedbackJson }),
                                            storage: { 'fw-grant': cfg.grant, 'fw-onboard-step': '2' } });
        await p.fill('#a1', line1); await p.fill('#city', 'Jasper');
        await p.fill('#state', 'GA'); await p.fill('#zip', '30143');
        await p.click('#go2');
        await p.waitForTimeout(500);
        const shown = await visible(p);
        const posts = p.__posts.filter((x) => x.url.includes('/api/feedback'));
        const id = posts.length ? posts[posts.length - 1].body && posts[posts.length - 1].body.id : null;
        const errs = p.__errors.slice();
        await p.__ctx.close();
        return { shown: shown.join(), id, errs };
      };

      const first  = await submit('123 Example Rd', { stored: 1 });
      const replay = await submit('123 Example Rd', { stored: 0, duplicate: true });
      const fixed  = await submit('456 Corrected Ave', { stored: 1 });

      record(8, 'replay of identical text → {stored:0,duplicate:true} accepted → s3, same record id',
             replay.shown === 's3' && replay.id !== null && replay.id === first.id,
             `screen=${replay.shown||'none'} firstId=${first.id} replayId=${replay.id}`);

      record(9, 'a CORRECTED address produces a DIFFERENT record id (not a silent no-op)',
             fixed.id !== null && first.id !== null && fixed.id !== first.id && fixed.shown === 's3',
             `firstId=${first.id} correctedId=${fixed.id} screen=${fixed.shown||'none'}`);
    }

    // ─── PATH 10 — a 200 is not proof of a write
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI, { status: 200, json: { stored: 0 } }),
                                          storage: { 'fw-grant': cfg.grant, 'fw-onboard-step': '2' } });
      await p.fill('#a1', '789 Nowhere Ln'); await p.fill('#city', 'Jasper'); await p.fill('#state', 'GA'); await p.fill('#zip', '30143');
      await p.click('#go2'); await p.waitForTimeout(500);
      const shown = await visible(p);
      const st = await p.evaluate(() => ({
        troubleHidden: document.getElementById('trouble').hidden,
        troubleText: (document.getElementById('trouble').textContent || '').trim(),
        disabled: document.getElementById('go2').disabled,
      }));
      record(10, '{stored:0} without duplicate → stays on s2, retry copy, button re-enabled',
             shown.join() === 's2' && !st.troubleHidden && !st.disabled,
             `screen=${shown.join()||'none'} buttonDisabled=${st.disabled} trouble="${st.troubleText.slice(0, 60)}"`);
      await p.__ctx.close();
    }

    // ─── PATH 11 — storage cleared AFTER the link was consumed. The link stripped ?g, so there is no way back.
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI), url: DOC_URL + '?g=' + cfg.grant });
      await p.evaluate(() => { try { localStorage.clear(); } catch (e) {} });
      await p.reload({ waitUntil: 'load' });
      await p.waitForTimeout(cfg.settleMs || 900);
      const shown = await visible(p);
      const html = await nolinkHTML(p);
      record(11, 'storage cleared after the link was consumed → resolves to a STATED outcome',
             shown.join() === 's-nolink', `screen=${shown.join()||'none'} copy="${(html||'').slice(0,70)}"`,
             { stated: 's-nolink (the bad-link screen)', note: 'what the RIGHT outcome is, is a content call — not asserted here' });
      await p.__ctx.close();
    }

    // ─── PATH 12 — a new device, link re-tapped
    {
      const p = await openPage(browser, { body: DOC, api: api(OK_WHOAMI), url: DOC_URL + '?g=' + cfg.grant });
      const shown = await visible(p);
      record(12, 'new device, original link re-tapped → s1', shown.join() === 's1', `screen=${shown.join()||'none'}`);
      await p.__ctx.close();
    }
  } catch (e) {
    record(0, 'runner fault', false, String(e && e.message || e));
  }

  await browser.close();
  console.log(JSON.stringify({ results }, null, 2));
})();
